lib/scholar/cluster/optics.ex

```diff
@@ -0,0 +1,257 @@
+defmodule Scholar.Cluster.OPTICS do
+  @moduledoc """
+  OPTICS (Ordering Points To Identify the Clustering Structure) is an algorithm
+  for finding density-based clusters in spatial data. It is closely related
+  to DBSCAN, finds core sample of high density and expands clusters from them. 
+  Unlike DBSCAN, keeps cluster hierarchy for a variable neighborhood radius. 
+  Clusters are then extracted using a DBSCAN-like method.
+  """
+  import Nx.Defn
+  require Nx
+
+  opts = [
+    min_samples: [
+      default: 5,
+      type: :pos_integer,
+      doc: """
+      The number of samples in a neighborhood for a point to be considered as a core point.
+      """
+    ],
+    max_eps: [
+      default: Nx.Constants.infinity(),
```

Comment:
I wouldn't set the default value here, rather in `fit/2`. I think it is better checking if `opts[:max_eps]` is `nil` and then setting it to `Nx.Constanst.infinity(Nx.type(x))`.

Reply:
It can be problematic, in edge case where `x` tensor of type {:s, 64} the following code will fail: 
```
x = Nx.tensor([[1, 2], [2, 5], [3, 6], [8, 7], [8, 8], [7, 3]])
Nx.Constants.infinity(Nx.type(x))
```

with error
```
** (FunctionClauseError) no function clause matching in Nx.Type.infinity_binary/1    
    
    The following arguments were given to Nx.Type.infinity_binary/1:
    
        # 1
        {:s, 64}
    
    Attempted function clauses (showing 4 out of 4):
    
        def infinity_binary({:bf, 16})
        def infinity_binary({:f, 16})
        def infinity_binary({:f, 32})
        def infinity_binary({:f, 64})
    
    (nx 0.7.3) lib/nx/type.ex:120: Nx.Type.infinity_binary/1
    (nx 0.7.3) lib/nx/constants.ex:96: Nx.Constants.infinity/2
    #cell:y6u577gyyh6jwkmd:2: (file)
```
so it seems that it will only work if `x` is float type tensor. 

Similar problem with hardcoded float type `reachability`:
`reachability = Nx.broadcast(Nx.Constants.max_finite({:f, 32}), {n_samples})`
I can't set its type to `Nx.type(x)` like this:
`reachability = Nx.broadcast(Nx.Constants.max_finite(Nx.type(x), {n_samples})`
because it will break each time when `Nx.type(x)` returns something different than float type - any of `{:bf, 16}`, `{:f, 16}`, `{:f, 32}`, `{:f, 64}`. `reachability` tensor should always be float type.

Reply:
You can use `Scholar.Shared.to_float_type/1`. You should convert `x` to a floating-point tensor using `Scholar.Shared.to_float` anyway.